broadinstitute/single_cell_portal_core
# Addressing CPU load in syncing large file counts (SCP-5295)

#### BACKGROUND
A recent production outage was found to have been caused by a user attempting to sync a study with 32K+ files in their workspace bucket.  Since sync will attempt to process all files at once - and will construct a map of folders/file types in memory - this caused the CPU to gradually increase over a span of tens of minutes.  Unfortunately, due to the long-running nature of the request, it eventually failed and caused the user to retry the operation repeatedly.  This eventually consumed all available CPU and caused the load balancer to block incoming requests after nginx responded `502`.

#### CHANGES
This update now limits all sync actions to 1000 files, and adds in a minimal pagination feature to notify the user how many remaining files there are, and gives them a link to process the next batch of 1000.  This batched sync will also only process workspace permissions on the first request, and all downstream calls only process files, not the workspace ACL.  This new functionality has been implemented in both the standard Rails MVC & API portions of the application.

Example notification of remaining files (with artificial batch size of 30)
<img width="1413" alt="Screenshot 2023-09-12 at 2 57 28 PM" src="https://github.com/broadinstitute/single_cell_portal_core/assets/729968/309996a5-07b1-4347-99e7-3d3809bb0fe8">


#### MANUAL TESTING
1. Boot as normal and sign in
2. Create a new study through the UI, then once you see the upload wizard, return to the 'My Studies' page
3. In a separate Rails console session, run the following block to seed the workspace bucket with files to process via sync:
```
study = Study.last
client = FireCloudClient.new
bucket = client.get_workspace_bucket(study.bucket_id)
file = File.open('test/test_data/metadata_example.txt')
puts "pushing #{File.basename(file)} to #{study.bucket_id}"
bucket.create_file file, File.basename(file)
Dir.chdir('tmp')
directory = 'file_type'
Dir.mkdir(directory)
50.times do
  filename = "#{directory}/#{SecureRandom.uuid}.csv"
  FileUtils.touch filename
  local = File.open(filename)
  puts "pushing #{filename} to #{study.bucket_id}"
  bucket.create_file local, filename
end
```
4. In `app/lib/study_sync_service.rb`, line 5, set `BATCH_SIZE` to `30` for testing
5. Back in the UI, find the entry for your new study, click the context menu on the right and select 'Sync workspace files'
6. Once the sync page loads, confirm you see the same message as above at the top of the page
7. Note the `DirectoryListing` for the `file_type` folder has `30` files
8. Click the link to `Process next batch`
9. Once the page loads, note you have a form for `metadata_example.txt`, and the file count for `file_type` has increased to `50`
